database/src/service/write.rs

```diff
@@ -141,10 +146,39 @@ impl DatabaseWriter {
             };
 
             // Map [`Request`]'s to specific database functions.
-            match request {
-                WriteRequest::Example1 => self.example_handler_1(response_sender),
-                WriteRequest::Example2(x) => self.example_handler_2(response_sender, x),
-                WriteRequest::Example3(x) => self.example_handler_3(response_sender, x),
+            // TODO: will there be more than 1 write request?
+            // this won't have to be an enum.
+            loop {
+                // TODO: Instead of reacting to resize errors, we should
+                // be proactively resizing each request. Else, we must
+                // `.clone()` the below block as the next `loop` retry
+                // must need access to it as well.
+                //
+                // If we had a local stack `usize` holding our
+                // current memory map bytes, we could calculate how
+                // much space we have, and if we need to proactively
+                // resize before writing some data to avoid a `.clone()`.
+
+                let response = match request {
+                    WriteRequest::WriteBlock(block) => write_block(&self.env, block),
+                };
+
+                match response {
+                    // TODO: what do we do if this errors?
+                    Ok(response) => response_sender.send(Ok(response)).unwrap(),
+                    Err(RuntimeError::ResizeNeeded) => {
+                        if request_retry_counter > REQUEST_RETRY_LIMIT {
+                            break;
+                        }
+
+                        request_retry_counter += 1;
+                        self.resize_map();
+                        continue;
+                    }
+                    Err(error) => response_sender.send(Err(error)).unwrap(),
+                }
+
+                break;
             }
```

Comment:
@Boog900 How do you think we should do retries on `WriteRequest`'s?

The current method of _reacting_ to resize errors instead of _proactively_ resizing comes with the downside in that we must `clone()` the data we use since we may not know if a resize is needed and the next `loop` iteration needs the data.

I propose we keep `RuntimeError::ResizeNeeded` but `DatabaseWriter` keeps track of the memory map byte size in some local `usize` counter, which we do book-keeping on each request:

1. `WriteRequest` received
2. Calculate how many bytes it'll write (with some buffer in-case?)
3. Check if our `current_memory_map_bytes: usize` needs to be resized
4. Resize if needed
5. Call write function
6. Panic if it errors with `ResizeNeeded` as we just handled that

Similar to: https://github.com/monero-project/monero/blob/c8214782fb2a769c57382a999eaf099691c836e7/src/blockchain_db/lmdb/db_lmdb.cpp#L665

Reply:
Nvm... ignore this.

Our `DatabaseRw::put` takes `&Key` not `Key`, so our write function will just take a reference, and we can be reactive when handling resize errors.